### PR TITLE
Adds fix for empty optional fields.

### PR DESCRIPTION
**Pull request by crwallace**

### CODE REVIEW

File: src/client/components/HealthCareApp.jsx [hunk 1]
```diff
@@ -29,15 +29,15 @@ class HealthCareApp extends React.Component {
     this.handleContinue = this.handleContinue.bind(this);
     this.handleSubmit = this.handleSubmit.bind(this);
     this.getUrl = this.getUrl.bind(this);
-    this.redirect = this.redirect.bind(this);
+    // this.redirect = this.redirect.bind(this);
```
Comment (awong-dev) [inline src/client/components/HealthCareApp.jsx]: delete?

File: src/client/components/HealthCareApp.jsx [hunk 2]
```diff
@@ -29,15 +29,15 @@ class HealthCareApp extends React.Component {
...
   }
 
   componentWillMount() {
-    this.redirect();
+    // this.redirect();
```
Comment (awong-dev) [inline src/client/components/HealthCareApp.jsx]: delete?

File: src/client/components/HealthCareApp.jsx [hunk 3]
```diff
@@ -29,15 +29,15 @@ class HealthCareApp extends React.Component {
...
   }
 
   componentDidUpdate() {
-    window.addEventListener('hashchange', this.redirect());
+    // window.addEventListener('hashchange', this.redirect());
```
Comment (awong-dev) [inline src/client/components/HealthCareApp.jsx]: delete?

File: src/client/components/HealthCareApp.jsx [hunk 4]
```diff
@@ -75,26 +75,26 @@ class HealthCareApp extends React.Component {
...
-    }
-
-    for (let j = 0; j < previousPaths.length; j++) {
-      if (this.props.uiState.sections[previousPaths[j]].complete === false) {
-        this.context.router.push('/introduction');
-        break;
-      }
-    }
-  }
+  // redirect() {
```
Comment (awong-dev) [inline src/client/components/HealthCareApp.jsx]: Just delete instead of commenting out.

File: src/common/schema/application.js [hunk 5]
```diff
@@ -54,10 +54,21 @@ module.exports = {
         'zipcode',
       ]
     },
-    date: {
+    requiredDate: {
       format: 'date',
       type: 'string'
     },
+    optionalDate: {
```
Comment (awong-dev) [inline src/common/schema/application.js]: Is this needed anymore if we know to remove the field at the client side before it's sent to the server?

File: src/common/schema/application.js [hunk 6]
```diff
@@ -71,7 +82,14 @@ module.exports = {
           type: 'string'
         },
         suffix: {
-          'enum': options.suffixes
+          oneOf: [
```
Comment (awong-dev) [inline src/common/schema/application.js]: Same here...can we just do this in the client side?

File: src/client/components/HealthCareApp.jsx (hunk 2 above)
Comment (crwallace) [inline src/client/components/HealthCareApp.jsx]: @akainic asked me to just comment it out for now.

File: src/common/schema/application.js [hunk 7]
```diff
@@ -167,7 +177,7 @@ module.exports = {
     },
     email: {
       type: 'string',
-      format: 'email'
+      pattern: '^(([a-zA-Z]|[0-9])|([-]|[_]|[.]))+[@](([a-zA-Z0-9])|([-])){2,63}[.](([a-zA-Z0-9]){2,63})+$|^$'
```
Comment (awong-dev) [inline src/common/schema/application.js]: Can we document where this regexp comes from in a comment?

File: src/common/veteran.js [hunk 8]
```diff
@@ -675,6 +675,22 @@ function veteranToApplication(veteran) {
...
+      // Optional Suffix fields
+      case 'veteranFullName':
+        if (value.suffix.value === '') { return undefined; }
+        break;
+
+      // Optional Date fields
+      case 'spouseDateOfBirth':
+      case 'dateOfMarriage':
+      case 'medicarePartAEffectiveDate':
+        if (value.day.value === '') { return undefined; }
```
Comment (awong-dev) [inline src/common/veteran.js]: Should this also test month/year to ensure they aren't empty?

Comment (awong-dev) [inline src/common/veteran.js]: Style nit: don't put all of the if statement on one line?

File: src/common/veteran.js [hunk 9]
```diff
@@ -675,6 +675,22 @@ function veteranToApplication(veteran) {
...
+        break;
+
+      // Optional SSN fields
```
Comment (awong-dev) [inline src/common/veteran.js]: You can combine the Optional Suffix fields and Optional SSN fields.

The grouping here makes the most sense to be if they are done by "type" instead of semantic value.  Basically, we have things that look like dates, and things that look like strings.

File: src/common/veteran.js [hunk 10]
```diff
@@ -675,6 +675,29 @@ function veteranToApplication(veteran) {
...
+      // Optional String fields
+      case 'spouseSocialSecurityNumber':
+        if (value.value === '') {
+          return undefined;
+        }
+        break;
+
+      case 'veteranFullName':
+      case 'spouseFullName':
+        if (value.suffix.value === '') {
```
Comment (akainic) [inline src/common/veteran.js]: Why only checking for `suffix` being blank and not also `middleName`?

File: src/common/schema/application.js [hunk 11]
```diff
@@ -128,9 +128,19 @@ module.exports = {
...
+        }, {
+          type: 'string',
+          pattern: '^$'
+        }
+      ]
+    },
     phone: {
       type: 'string',
-      pattern: '^[0-9]{10}$'
+      pattern: '^[0-9]{10}$|^$'
```
Comment (markolson) [inline src/common/schema/application.js]: Do you want to the phone number type itself is always allowed to be blank, or do we care more about specific uses of it (i.e. homePhone must exist, but spousePhone can be blank)?

Comment (crwallace) [inline src/common/schema/application.js]: None of the phone numbers are required fields so no real need for that.

Comment (awong-dev) [inline src/common/schema/application.js]: The pattern at the API layer should not accept blank.

The changes to `phone` and `state` should likely be reverted in favor of removing them from the object before it is even sent in the server.

**Comment by awong-dev** [comment]
LGTM. Let's get this in...once I fix CI.... grrrrrrr
